File: python/mirage/commands/local_audio/disk/tail_wav.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import AsyncIterator

from mirage.cache.index import IndexCacheStore
from mirage.commands.local_audio.utils import metadata, transcribe
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.disk.glob import resolve_glob
from mirage.core.disk.read import read_bytes
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


@command("tail", resource="disk", spec=SPECS["tail"], filetype=".wav")
async def tail_wav(
    accessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    n: str | None = None,
    c: str | None = None,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if not paths:
        raise ValueError("tail: missing operand")
    paths = await resolve_glob(accessor, paths, index)
    if c is not None:
        return None, IOResult(
            exit_code=1,
            stderr=b"tail: -c not supported for audio files",
        )
    try:
        seconds = int(n) if n is not None else 10
        raw = await read_bytes(accessor, paths[0])
        meta = metadata(raw)
        duration = meta.get("duration", 0) or 0
        start = max(0.0, duration - seconds)
        stream = transcribe(raw, start_sec=start)
        return stream, IOResult(reads={paths[0].strip_prefix: raw},
                                cache=[paths[0].strip_prefix])
    except Exception as e:
        return None, IOResult(
            exit_code=1,
            stderr=f"tail: {paths[0].original}: failed to read as wav: {e}".
            encode(),
        )

File: typescript/packages/node/src/commands/local_audio/ram/tail_wav.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import {
  IOResult,
  ResourceName,
  command,
  specOf,
  type ByteSource,
  type CommandFnResult,
  type CommandOpts,
  type PathSpec,
  type RAMAccessor,
} from '@struktoai/mirage-core'
import { metadata, transcribe } from '../utils.ts'
import { readRam } from './read.ts'

const ENC = new TextEncoder()

async function tailWavCommand(
  accessor: RAMAccessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  if (paths.length === 0) {
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('tail: missing operand\n') })]
  }
  const first = paths[0]
  if (first === undefined) return [null, new IOResult()]
  if (typeof opts.flags.c === 'string') {
    return [
      null,
      new IOResult({ exitCode: 1, stderr: ENC.encode('tail: -c not supported for audio files\n') }),
    ]
  }
  const n = typeof opts.flags.n === 'string' ? Number.parseInt(opts.flags.n, 10) : 10
  try {
    const raw = readRam(accessor, first)
    const meta = await metadata(raw)
    const duration = meta.duration ?? 0
    const start = Math.max(0, duration - n)
    const out: ByteSource = transcribe(raw, start)
    return [out, new IOResult({ reads: { [first.stripPrefix]: raw }, cache: [first.stripPrefix] })]
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err)
    return [
      null,
      new IOResult({
        exitCode: 1,
        stderr: ENC.encode(`tail: ${first.original}: failed to read as wav: ${msg}\n`),
      }),
    ]
  }
}

export const RAM_LOCAL_AUDIO_TAIL_WAV = command({
  name: 'tail',
  resource: ResourceName.RAM,
  spec: specOf('tail'),
  filetype: '.wav',
  fn: tailWavCommand,
})
